include/swift/Basic/OwnedString.h

```diff
@@ -22,117 +22,89 @@
 
 #include "llvm/ADT/IntrusiveRefCntPtr.h"
 #include "llvm/ADT/StringRef.h"
+#include "llvm/Support/TrailingObjects.h"
 
 using llvm::StringRef;
 
 namespace swift {
 
-enum class StringOwnership {
-  /// An OwnedString holds a weak reference to the underlying string storage
-  /// and will never attempt to free it.
-  Unowned,
-
-  /// An OwnedString has its own copy of the underlying string storage and
-  /// will free the storage upon its destruction.
-  Copied,
-};
-
 /// Holds a string - either statically allocated or dynamically allocated
 /// and owned by this type.
 class OwnedString {
-  const char *Data;
-  size_t Length;
-  StringOwnership Ownership = StringOwnership::Unowned;
-
-  void release() {
-    if (Ownership == StringOwnership::Copied)
-      free(const_cast<char *>(Data));
-  }
+  /// An owner that keeps the buffer of a ref counted \c OwnedString alive.
+  class TextOwner final : public llvm::ThreadSafeRefCountedBase<TextOwner>,
+                          public llvm::TrailingObjects<TextOwner, char> {
+    TextOwner(StringRef Text) {
+      std::uninitialized_copy(Text.begin(), Text.end(),
+                              getTrailingObjects<char>());
+    }
 
-  void initialize(const char* Data, size_t Length, StringOwnership Ownership) {
-    this->Length = Length;
-    this->Ownership = Ownership;
-    if (Ownership == StringOwnership::Copied && Data) {
-      char *substring = static_cast<char *>(malloc(Length + 1));
-      assert(substring && "expected successful malloc of copy");
+  public:
+    static TextOwner *make(StringRef Text) {
+      auto size = totalSizeToAlloc<char>(Text.size());
+      void *data = ::operator new(size);
+      return new (data) TextOwner(Text);
+    }
 
-      memcpy(substring, Data, Length);
-      substring[Length] = '\0';
+    const char *getText() const { return getTrailingObjects<char>(); }
```

Comment:
Is there any specific reasons to put `char*` as a trailing object instead of a regular member?

Reply:
It allows us to allocate the storage with a single allocation instead of two which should be less overhead.

Reply:
ah, OK. The patch LGTM!